ouroboros-network/src/Ouroboros/Network/ErrorPolicy.hs

```diff
@@ -0,0 +1,313 @@
+{-# LANGUAGE DeriveFunctor         #-}
+{-# LANGUAGE GADTs                 #-}
+{-# LANGUAGE FlexibleContexts      #-}
+{-# LANGUAGE NamedFieldPuns        #-}
+{-# LANGUAGE ScopedTypeVariables   #-}
+
+-- | Error policies, and integration with 'SuspendDecision'-semigroup action on
+-- 'PeerState'.
+--
+module Ouroboros.Network.ErrorPolicy
+  ( ErrorPolicies (..)
+  , ConnectionException (..)
+  , ApplicationException (..)
+  , nullErrorPolicies
+  , ErrorPolicy (..)
+  , evalErrorPolicy
+  , evalErrorPolicies
+  , CompleteApplication
+  , Result (..)
+  , completeApplicationTx
+
+  -- * Traces
+  , ErrorPolicyTrace (..)
+  , traceErrorPolicy
+  , WithAddr (..)
+
+  -- * Re-exports of PeerState
+  , PeerStates
+  , SuspendDecision (..)
+  ) where
+
+import           Control.Exception (Exception, SomeException (..))
+import           Data.List.NonEmpty (NonEmpty (..))
+import           Data.Foldable (traverse_)
+import qualified Data.Map.Strict as Map
+import           Data.Maybe (fromMaybe, mapMaybe)
+import           Data.Semigroup (sconcat)
+import           Data.Set (Set)
+import qualified Data.Set as Set
+import           Data.Typeable ( Proxy (..)
+                               , gcast
+                               , tyConName
+                               , typeRepTyCon
+                               , typeRep
+                               )
+import           Text.Printf
+
+import           Control.Monad.Class.MonadAsync
+import           Control.Monad.Class.MonadSTM
+import           Control.Monad.Class.MonadTime
+import           Control.Tracer
+
+import           Data.Semigroup.Action
+
+import           Ouroboros.Network.Subscription.PeerState
+
+data ErrorPolicy e where
+     ErrorPolicy :: forall e err.
```

Comment:
`e` is a type parameter but also `forall`-quantified here, is that necessary?

Reply:
If that somehow is necessary, please put `err` first, because I'd like to use type application to instantiate `err`.

Reply:
Yes, `forall` for `e` is needed, but I will put `err` first.

Reply:
I see. Ah, it's because of the `forall`, just like for functions, when there's a rank-1 `forall`, all type variables need to be explicitly quantified.

I have just checked locally that the `forall` isn't even needed. It even expects `err` as  its first type argument! I prefer the explicit `forall` here, though, to signal that there is an explicit quantification.